### ISSUE TITLE
Ability to filter samplings based on input of ExplorationTask

### ISSUE THREAD

**Issue by schuhschuh** [openmole/openmole#69]
I have a workflow where I want to explore all pairs of images, but using two chained ExplorationTask's instead of a single one so I can first aggregate the inner "loop" before the outer one.

For example, the following is currently not possible

```scala
val i = Val[Int]
val j = Val[Int]
val k = Val[Int]

val exploration1 = ExplorationTask(i in (1 to 10))
val exploration2 = ExplorationTask((j in (1 to 10)) filter "i != j") set (inputs += i, outputs += i)
val task = ScalaTask("val k = i + j") set (inputs += (i, j), outputs += k)
val zero = EmptyTask() set (inputs += (i, j), outputs += k, k := 0)

val ex = exploration1 -< exploration2 -< task >- average
```

One option might be to use a Switch as follows:
```scala
val exploration2 = ExplorationTask(j in (1 to 10)) set (inputs += i, outputs += i)
val ex = exploration1 -< exploration2 -< Switch(
  Case("i != j", task),
  Case("i == j", zero)
) >- average
```
However, given that in my case I am not dealing with integers, but files instead, the "average" task would presumably throw an exception if it receives an invalid File object.

Another workaround could be to use a single ExplorationTask, but then I cannot easily average the results for each "i" in parallel easily. This would need a second ExplorationTask after the aggregation:
```scala
val i = Val[Int]
val j = Val[Int]
val k = Val[Int]
val l = Val[Int]

val exploration1 = ExplorationTask(((i in (1 to 10)) x (j in (1 to 10))) filter "i != j")
val exploration2 = Capsule(ExplorationTask((l in (1 to 10)), strainer = true)
val task = ScalaTask("val k = i + j") set (inputs += (i, j), outputs += (i, k))

val ex = exploration1 -< task >- exploration2 -< average
```
where the "average" task would only consider those elements in the aggregated "k" array with a corresponding "i" value equal to "l".

None of these alternative workflows are ideal IMHO.

An alternative that would be fine for me is:
```scala
val i = Val[Int]
val j = Val[Int]
val k = Val[Int]
val m = Val[Double]

val exploration1 = ExplorationTask(i in (1 to 10))
val exploration2 = ExplorationTask(j in (1 to 10)) set (inputs += i, outputs += i)
val task = ScalaTask("val k = i + j") set (inputs += (i, j), outputs += k)
val average = ScalaTask("val m = input.k.sum / input.k.size") set (inputs += k.toArray, outputs += m)

val ex = exploration1 -< exploration2 -< (task when "i != j") >- (average hook ToStringHook())
```
but first of all, I am not sure if the "average" task will ever execute when one of the parallel streams was terminated by the "when" (not like a Skip), and secondly this gives me the following strange error:
```
OpenMOLE><console>:17: warning: comparing values of types Int and Array[Int] using `!=' will always yield true
                  i != j
```

I found a working example using "when". First of, my concern that pruning the "i == j" branch would result  in "average" not being executed is unjustified. It works just fine. In order to get rid of above error, I have to add a dummy task in between the exploration transition and the task with the "when" condition on it.

Here a full working version:
```scala
val i = Val[Int]
val j = Val[Int]
val k = Val[Int]

val forAllI = ExplorationTask(i in (1 to 10))
val forAllJ = Capsule(ExplorationTask(j in (1 to 10)), strainer = true)
val dummy = Capsule(EmptyTask(), strainer = true)
val multiply = ScalaTask("val k = i * j") set (inputs += (i, j), outputs += k)
val print = (EmptyTask() set (inputs += k.toArray, outputs += k.toArray)) hook ToStringHook()

val ex = forAllI -< forAllJ -< dummy -- (multiply when "i != j") >- print start
```

Any chance that the need for the "dummy" strainer just to have a normal transition "--" before the task with the "when" condition on it can be removed ? It isn't very intuitive...

**Comment by romainreuillon** [comment]
Hi Andreas,

first I don't catch why this is not possible ?

val  i  =  Val[Int]
val  j  =  Val[Int]
val  k  =  Val[Int]

val  exploration1  =  ExplorationTask(i in (1  to10))
val  exploration2  =  ExplorationTask((j in (1  to10)) filter"i != j") set (inputs+=  i, outputs+=  i)
val  task  =  ScalaTask("val k = i + j") set (inputs+=  (i, j), outputs+=  k)
val  average=  ScalaTask("k = k.sum.toDouble / k.size") set (inputs+=  k.toArray, outputs+=  k)

val  ex  =  exploration1-<  exploration2-<  task>-  average

For the other pb:

|OpenMOLE><console>:17: warning: comparing values of types Int and Array[Int] using `!=' will always yield true
                   i != j|

This warning occurs because the condition is tested withing the begining 
context of the exploration transition. Alternatively it could be 
evaluated before submitting job in the exploration transition, I am 
unsure what make the most sens right now. What do you think ?

Romain

**Comment by schuhschuh** [comment]
When I try this example (slight modification below to fix syntax screwed up by mail client and simply print the result of the aggregation):

```scala
val i = Val[Int]
val j = Val[Int]
val k = Val[Int]

val exploration1 = ExplorationTask(i in (1 to 10))
val exploration2 = ExplorationTask((j in (1 to 10)) filter "i != j") set (inputs += i, outputs += i)
val task         = ScalaTask("val k = i * j") set (inputs += (i, j), outputs += k)
val print        = EmptyTask() set (inputs += k.toArray, outputs += k.toArray) hook ToStringHook()

val ex = exploration1 -< exploration2 -< task >- print
```

I get the following compilation error at runtime:
```
OpenMOLE><console>:17: error: not found: value i
                  i != j
                  ^
Jun 05, 2015 2:34:22 PM org.openmole.core.workflow.mole.SubMoleExecution finalState
SEVERE: Error in user job execution for capsule explorationTask11, job state is FAILED.
org.openmole.core.exception.InternalProcessingError: Error for context values in explorationTask11 {i=10, oMSeed=7752694624879150030}
        at org.openmole.core.workflow.tools.InputOutputCheck$class.perform(InputOutputCheck.scala:94)
        at org.openmole.core.workflow.task.ExplorationTask.perform(ExplorationTask.scala:55)
        at org.openmole.core.workflow.task.Task$class.perform(Task.scala:47)
        at org.openmole.core.workflow.task.ExplorationTask.perform(ExplorationTask.scala:55)
        at org.openmole.core.workflow.job.MoleJob.perform(MoleJob.scala:99)
        at org.openmole.core.workflow.execution.local.LocalExecutor$$anonfun$1$$anonfun$apply$1.apply(LocalExecutor.scala:70)
        at org.openmole.core.workflow.execution.local.LocalExecutor$$anonfun$1$$anonfun$apply$1.apply(LocalExecutor.scala:62)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at org.openmole.core.workflow.execution.local.LocalExecutor$$anonfun$1.apply(LocalExecutor.scala:62)
        at org.openmole.core.workflow.execution.local.LocalExecutor$$anonfun$1.apply(LocalExecutor.scala:59)
        at org.openmole.core.workflow.execution.local.LocalExecutor.withRedirectedOutput(LocalExecutor.scala:113)
        at org.openmole.core.workflow.execution.local.LocalExecutor.run(LocalExecutor.scala:59)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.openmole.core.exception.InternalProcessingError: Error compiling import org.openmole.core.workflow.tools.CodeTool._

(_input_value_context: org.openmole.core.workflow.data.Context, _input_value_RNGProvider: org.openmole.core.workflow.data.RandomProvider) => {
    object input {
      var j = _input_value_context("j").asInstanceOf[Int]
    }
    import input._
    implicit lazy val oMRNG: util.Random = _input_value_RNGProvider()
    i != j
    
}

        at org.openmole.core.workflow.tools.ScalaCompilation$$anonfun$compile$1.apply(ScalaCompilation.scala:42)
        at scala.util.Try$.apply(Try.scala:191)
        at org.openmole.core.workflow.tools.ScalaCompilation$class.compile(ScalaCompilation.scala:38)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$$anon$1.compile(ScalaCompilation.scala:61)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$class.function(ScalaCompilation.scala:136)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$$anon$1.function(ScalaCompilation.scala:61)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$$anonfun$compiled$1.apply(ScalaCompilation.scala:181)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$$anonfun$compiled$1.apply(ScalaCompilation.scala:181)
        at scala.collection.mutable.MapLike$class.getOrElseUpdate(MapLike.scala:194)
        at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:80)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$class.compiled(ScalaCompilation.scala:179)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$$anon$1.compiled(ScalaCompilation.scala:61)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$class.compiled(ScalaCompilation.scala:194)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$$anon$1.compiled(ScalaCompilation.scala:61)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$class.run(ScalaCompilation.scala:197)
        at org.openmole.core.workflow.tools.ScalaWrappedCompilation$$anon$1.run(ScalaCompilation.scala:61)
        at org.openmole.plugin.sampling.combine.ScalaSamplingFilter.apply(FilteredSampling.scala:54)
        at org.openmole.plugin.sampling.combine.FilteredSampling$$anonfun$build$1$$anonfun$apply$1.apply(FilteredSampling.scala:39)
        at org.openmole.plugin.sampling.combine.FilteredSampling$$anonfun$build$1$$anonfun$apply$1.apply(FilteredSampling.scala:39)
        at scala.collection.IndexedSeqOptimized$class.prefixLengthImpl(IndexedSeqOptimized.scala:38)
        at scala.collection.IndexedSeqOptimized$class.exists(IndexedSeqOptimized.scala:46)
        at scala.collection.mutable.WrappedArray.exists(WrappedArray.scala:35)
        at org.openmole.plugin.sampling.combine.FilteredSampling$$anonfun$build$1.apply(FilteredSampling.scala:39)
        at org.openmole.plugin.sampling.combine.FilteredSampling$$anonfun$build$1.apply(FilteredSampling.scala:39)
        at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:415)
        at scala.collection.Iterator$class.foreach(Iterator.scala:750)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1202)
        at org.openmole.core.workflow.task.ExplorationTask.process(ExplorationTask.scala:62)
        at org.openmole.core.workflow.task.Task$$anonfun$perform$1.apply(Task.scala:47)
        at org.openmole.core.workflow.task.Task$$anonfun$perform$1.apply(Task.scala:47)
        at org.openmole.core.workflow.tools.InputOutputCheck$class.perform(InputOutputCheck.scala:91)
        ... 12 more
Caused by: org.openmole.core.exception.UserBadDataError: Error while compiling:
not found: value i
                  i != j
                  ^
on line 17
        at org.openmole.core.console.ScalaREPL.eval(ScalaREPL.scala:70)
        ... 43 more
```

**Comment by schuhschuh** [comment]
> For the other pb:
>
> |OpenMOLE><console>:17: warning: comparing values of types Int and Array[Int] using `!=' will always > yield true
>                  i != j|
>
> This warning occurs because the condition is tested within the beginning
context of the exploration transition. Alternatively it could be
evaluated before submitting job in the exploration transition, I am
unsure what make the most sense right now. What do you think ?

To me, the "when" condition, i.e, conditional transitions were always a bit confusing. The DSL syntax suggests to me that the condition is applied to a task rather than a transition, but the latter seems to be actually the case internally (thus, you call them "conditional transition" whereas I would treat them as "conditional task execution"). After all, I write
```scala
exploration -< (t2 when "true")
```
instead of
```scala
exploration (-< when "true") t2
```
(which is odd anyway... just for illustration).

Hence, I would expect that the condition is evaluated in the same context as the task on which it is placed on. Therefore, the condition should be evaluated once for each input that the task would receive and thus the input variables should not be an array of the to be explored values, but only a single one of these.

**Comment by schuhschuh** [comment]
BTW I am using the master branch at commit 7d9de0029ad5d6af333d2e6a0f7f885d4e2952d2.

**Comment by romainreuillon** [comment]
The error in filter is fixed by c7d62e58abaeaac8b389208ebc30c396138f0239.

For the condition on the exploration transition your point of view make sense but I have to think a bit more about that since it would provide a feature which would be really redundant with the filters and we would have no way to skip entirely the transition.

**Comment by schuhschuh** [comment]
Thanks, to apply the filter to the sampling directly was my first attempt in this scenario anyway.
